Core/Src/logger.c

```diff
@@ -0,0 +1,171 @@
+    curr_logger_block = 0;
+    curr_logger_page = 0;
+
+    logger_addr.block = curr_logger_block;
+    logger_addr.page = curr_logger_page;
+
+    buffer_pointer = 0;
+}
+
+void iris_log(const char *log_data, ...) {
+    uint8_t output_array[LOG_TOTAL_LENGTH];
+    memset(output_array, 0, LOG_TOTAL_LENGTH);
+
+    uint8_t header_buffer[LOG_HEADER_LENGTH];
+    uint8_t data_buffer[LOG_DATA_LENGTH];
+    uint8_t footer_buffer[LOG_FOOTER_LENGTH];
+
+    Iris_Timestamp timestamp;
+    get_rtc_time(&timestamp);
+
+    // Create header
+    snprintf(header_buffer, LOG_HEADER_LENGTH, "%d/%d/%d %d:%d:%d: ", timestamp.Day, timestamp.Month,
+             timestamp.Year, timestamp.Hour, timestamp.Minute, timestamp.Second);
+
+    // Create data
+    va_list arg;
+    va_start(arg, log_data);
+    vsnprintf(data_buffer, LOG_DATA_LENGTH, log_data, arg);
+
+    // Create footer
+    snprintf(footer_buffer, LOG_FOOTER_LENGTH, "\r\n");
+
+    // Combine footer
+    snprintf(output_array, LOG_TOTAL_LENGTH, "%s%s%s", header_buffer, data_buffer, footer_buffer);
+
+    _fill_buffer(output_array);
+}
+
+void _fill_buffer(uint8_t *data) {
+    uint8_t count = 0;
+
+    if (buffer_pointer >= PAGE_DATA_SIZE) {
+        clear_and_dump_buffer();
+    }
+
+    for (int i = buffer_pointer; i < (buffer_pointer + LOG_TOTAL_LENGTH); i++) {
+        logger_buffer[i] = data[count];
+        count++;
+    }
+
+    buffer_pointer += LOG_TOTAL_LENGTH;
+}
+
+int _logger_write() {
+    NAND_ReturnType ret = NAND_Page_Program(&logger_addr, PAGE_DATA_SIZE, logger_buffer);
+    if (ret != Ret_Success) {
+        DBG_PUT("Error: prog b %d p %d r %d\r\n", logger_addr.block, logger_addr.page, ret);
+        return ret;
+    }
+
+    DBG_PUT("prog b %d p %d r %d\r\n", logger_addr.block, logger_addr.page, ret);
+
+    curr_logger_page++;
+
+    if (curr_logger_page >= NUM_PAGES_PER_BLOCK) {
+        curr_logger_block = curr_logger_block ^ LOG_BLOCK_SWITCH_MASK;
+        curr_logger_page = 0;
+
+        logger_addr.block = curr_logger_block;
+
+        ret = NAND_Block_Erase(&logger_addr);
+        if (ret != Ret_Success) {
+            DBG_PUT("Erase block %d failed", logger_addr.block);
+            return ret;
+        }
+    }
+
+    logger_addr.page = curr_logger_page;
+    return 0;
+}
+
+int clear_and_dump_buffer() {
+    int ret;
+    ret = _logger_write();
+
+    if (ret != 0) {
+        return -1;
+    }
+
+    buffer_pointer = 0;
+    memset(logger_buffer, 0, PAGE_DATA_SIZE);
+
+    return 0;
+}
+
+int logger_clear() {
+    PhysicalAddrs addr = {0};
+
+    for (uint8_t blk = LOG_BLOCK_LOW; blk <= LOG_BLOCK_HIGH; blk++) {
+        addr.block = blk;
+        addr.page = 0;
+
+        NAND_ReturnType ret = NAND_Block_Erase(&addr);
+        if (ret != Ret_Success) {
+            DBG_PUT("Error in deleting block b %d p %d r %d\r\n", addr.block, addr.page, ret);
+            return ret;
+        }
+    }
+    return 0;
+}
+
+int read_from_block(uint8_t block, uint16_t page) {
```

Comment:
Is this a test function? It doesn't do anything with the data other than DBG_PUT it

Reply:
I have moved this to the nand testing directory